src/pe/relocation.rs

```diff
@@ -131,3 +308,364 @@ impl<'a> Iterator for Relocations<'a> {
+    pub fn blocks(&self) -> RelocationBlockIterator<'a> {
+        RelocationBlockIterator {
+            bytes: &self.bytes,
+            offset: 0,
+        }
+    }
+}
+
+/// Represents a PE (Portable Executable) base relocation block.
+///
+/// In the PE format, base relocations are organized into blocks, where each block corresponds
+/// to a contiguous range of memory that needs to be relocated. The relocations within a block
+/// specify offsets relative to the block's starting address.
+///
+/// Unlike COFF (Common Object File Format), PE base relocations are separated by blocks.
+/// Each block begins with a [`RelocationBlock`] structure that provides information about the
+/// relocation block, including its starting RVA (Relative Virtual Address) and size in bytes.
+///
+/// Each block ends with an "absolute" relocation entry, which acts as a marker to indicate
+/// the end of the block. This entry has no actual relocation effect.
+#[derive(Debug, Copy, Clone, PartialEq, Default)]
+pub struct RelocationBlock<'a> {
+    /// The starting RVA for the relocation block.
+    pub rva: u32,
+    /// The total size of the relocation block, in bytes.
+    pub size: u32,
+    /// The bytes of [`RelocationWord`] data within this block.
+    pub bytes: &'a [u8],
+}
+
+impl RelocationBlock<'_> {
+    /// Returns iterator for [`RelocationWord`]
+    pub fn words(&self) -> RelocationWordIterator {
+        RelocationWordIterator {
+            bytes: &self.bytes,
+            offset: 0,
+        }
+    }
+}
+
+/// Raw size of [`RelocationBlock`]
+pub const RELOCATION_BLOCK_SIZE: usize = 8;
+
+impl<'a> scroll::ctx::TryFromCtx<'a, scroll::Endian> for RelocationBlock<'a> {
+    type Error = crate::error::Error;
+
+    fn try_from_ctx(bytes: &'a [u8], ctx: scroll::Endian) -> Result<(Self, usize), Self::Error> {
+        let mut offset = 0;
+        let rva = bytes.gread_with::<u32>(&mut offset, ctx)?;
+        let size = bytes.gread_with::<u32>(&mut offset, ctx)?;
+        let bytes = &[];
+        Ok((Self { rva, size, bytes }, offset))
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct RelocationBlockIterator<'a> {
+    bytes: &'a [u8],
+    offset: usize,
+}
+
+impl<'a> Iterator for RelocationBlockIterator<'a> {
+    type Item = error::Result<RelocationBlock<'a>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.offset >= self.bytes.len() {
+            return None;
+        }
+
+        Some(
+            match self
+                .bytes
+                .gread_with::<RelocationBlock>(&mut self.offset, scroll::LE)
+            {
+                Ok(block) => {
+                    let word_size = block.size as usize - RELOCATION_BLOCK_SIZE;
+                    let word_bytes = &self.bytes[self.offset..self.offset + word_size];
+                    // "Block"s and "Word"s are entirely contiguous
+                    self.offset += word_size;
+                    Ok(RelocationBlock {
+                        bytes: word_bytes,
```

Comment:
i see, this gets rewritten later. hmmm, one thing i don't like is that temporarily, if someone were to use that raw TryFromCtx impl the bytes is empty, which feels "half-constructed" to me, which is generally a pattern i like to avoid.

however, not sure there is a better solution here, other than making it optional, which also feels a little laborious and still doesn't capture semantic meaning that well

Reply:
It's been quite a while since I wrote the code and I have no idea why did I chose to do this ugly.. anyway, I fixed that. It was trivial to just merge that construction into the `TryFromCtx for RelocationBlock`.